zerver/actions/message_edit.py

```diff
@@ -885,6 +886,77 @@ def user_info(um: UserMessage) -> Dict[str, Any]:
+    if (
+        realm.move_messages_within_stream_limit_seconds is None
+        and realm.move_messages_between_streams_limit_seconds is None
+    ):
+        # There is no time limit for editing both stream and topics, so we need
+        # not check further and moving message is allowed.
+        return
+
+    message_move_limit_buffer = 20
+
+    topic_edit_deadline_seconds = None
+    if topic_name is not None and realm.move_messages_within_stream_limit_seconds is not None:
+        # We set topic_edit_deadline_seconds only if topic is actually changed and there is
+        # some time limit to edit topic.
+        topic_edit_deadline_seconds = (
+            realm.move_messages_within_stream_limit_seconds + message_move_limit_buffer
+        )
+
+    stream_edit_deadline_seconds = None
+    if stream_id is not None and realm.move_messages_between_streams_limit_seconds is not None:
+        # We set stream_edit_deadline_seconds only if stream is actually changed and there is
+        # some time limit to edit stream.
+        stream_edit_deadline_seconds = (
+            realm.move_messages_between_streams_limit_seconds + message_move_limit_buffer
+        )
+
+    # If the code reaches here, then there is a time limit to edit at least one of topic or stream.
+    assert topic_edit_deadline_seconds is not None or stream_edit_deadline_seconds is not None
+
+    if topic_edit_deadline_seconds is not None and stream_edit_deadline_seconds is not None:
+        # When both stream and topic are changed, we consider the minimum of the two limits to
+        # make sure that we raise the error even when user cannot change one of topic or stream.
+        message_move_deadline_seconds = min(
+            topic_edit_deadline_seconds, stream_edit_deadline_seconds
+        )
+    elif topic_edit_deadline_seconds is not None:
+        message_move_deadline_seconds = topic_edit_deadline_seconds
+    elif stream_edit_deadline_seconds is not None:
+        message_move_deadline_seconds = stream_edit_deadline_seconds
+
+    allowed_messages_query = Q(
+        recipient_id=message.recipient_id,
+        subject__iexact=message.topic_name(),
+        date_sent__gt=timezone_now() - datetime.timedelta(seconds=message_move_deadline_seconds),
+    )
+    messages_allowed_to_move = (
+        Message.objects.filter(allowed_messages_query)
+        .order_by("date_sent")
+        .values_list("id", flat=True)
+    )
+    all_messages_in_topic = Message.objects.filter(
+        recipient_id=message.recipient_id, subject__iexact=message.topic_name()
+    ).values_list("id", flat=True)
+
+    if len(all_messages_in_topic) == len(messages_allowed_to_move):
+        # We return if all messages are allowed to move.
+        return
+
+    raise MessageMoveError(
+        first_message_allowed_to_move=messages_allowed_to_move[0],
+        total_messages_in_topic=len(all_messages_in_topic),
+        messages_allowed_to_move=len(messages_allowed_to_move),
+    )
```

Comment:
I have a few concerns with the logic here. One is that you're doing direct queries on `Message`, which don't check whether the acting user has permission to see all of those messages -- a potential security concern in a stream without shared history. 

The second is that we do a very similar set of queries inside `do_update_message`, being careful about this security issue, to determine what sort of notice to display from Notification Bot on whether the entire topic was moved or not; and probably it'd be good to avoid duplicate code/queries for that, which I suspect (but have not confirmed) we're doing here.

If we didn't need to think about the permissions issue, we could do this query much more efficiently with just filtering `messages_for_topic` for the oldest message in the topic, and checking whether or not it's older? It's possible we want to do that check to have a fast path for the case that it's going to be allowed, since that query is likely very fast, and fetching all the message IDs in a topic may not be?

Idunno, maybe that's too complicated, and we should just have two cases -- (1) the current user has full history access, where we can roughly what you have above, and (2) they don't, and we need to compute both `messages_allowed_to_move` and `all_messages_in_topic` via a query that filters on `UserMessage` too. 

(In any case, it was probably a good idea you put this in its own function given how complicated it's looking like it'll be!)

Reply:
> I am not sure whether we can just deduplicate code though by extracting some functions, but I think it would make code somewhat hard to read. In any case, I can try if we can do this.

Yeah, maybe it's fine to leave this as-is.